Fix usage of `amp_register_polyfills` hook

thelovekesh opened the pull request:
## Summary

This PR accounts for the following changes:
- Make changes to the Webpack config to fix the default export in `@wordpress/dom-ready` package.
- Remove `amp_register_polyfills` from the following services/classes:
    - PairedBrowsing -> This service has a conditional of `has_support` from `DependecySupport` which means it will work only on WP >= 5.6 so no polyfills are required. Also, this service has no dependency on React so we can keep it on our plugin.
    - ValidationCounts -> This service has a conditional of `has_support` from `DependecySupport` which means it will work only on WP >= 5.6 so no polyfills are required. Also, this service has no dependency on React so we can keep it on our plugin.
    - AMP_Template_Customizer -> We have disabled AMP customizer on WP < 5.6 so no polyfills required. Also, this service has no dependency on React so we can keep it on our plugin.
 - Disable AMP customizer on WP < 5.6
---

With this, we only have the following services left which use `amp_register_polyfills` only on their dedicated screens:
- OnboardingWizardSubmenuPage
- OptionsMenu
- SupportScreen

Fixes #7457

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

thelovekesh committed:
Update ValidationCounts service to register `amp_register_polyfills` hook when screen is AMP options menu section

thelovekesh committed:
Update `amp_register_polyfills` to register Polyfills only on older WP versions

westonruter commented on src/Admin/ValidationCounts.php:
```diff
@@ -89,8 +89,9 @@ public function register() {
 	public function enqueue_scripts() {
-		/** This action is documented in includes/class-amp-theme-support.php */
-		do_action( 'amp_register_polyfills' );
+		if ( $this->is_options_menu() ) {
```
Good idea.

westonruter replied:
Oh, actually maybe this won't work as expected? Let's say you're _not_ on the Settings screen. In this case, the polyfills won't be registered and the result will be possibly that the required scripts are not registered, right?

I think the way to address this is to remove the logic here and move it to `is_needed`, so that the logic runs if:

We have the expected version of React _or_ we're on one of the AMP settings screens.

thelovekesh replied:
> In this case, the polyfills won't be registered and the result will be possibly that the required scripts are not registered, right?

Since it has `'wp-api-fetch', 'wp-dom-ready', 'wp-i18n'` as dependency scripts while `wp_enqueue_script`, WP will enqueue them for it... right? I think WP prior to `5.6` already has these scripts' support and below `5.6` we are not exposing this service.

https://github.com/ampproject/amp-wp/blob/378522cfc969ed1d04abb1b0f1605838ac24f562/src/Admin/ValidationCounts.php#L78

thelovekesh replied:
> We have the expected version of React or we're on one of the AMP settings screens.

`amp-validation-counts.js` doesn't require React for its operation. Here is the dependecies for this:
```php
<?php return array('dependencies' => array('wp-api-fetch', 'wp-dom-ready', 'wp-i18n'), 'version' => '29b558177338ddf7e50f');
```

westonruter commented on src/Admin/ValidationCounts.php:
```diff
@@ -89,8 +89,9 @@ public function register() {
 	 * Enqueue admin assets.
 	 */
 	public function enqueue_scripts() {
-		/** This action is documented in includes/class-amp-theme-support.php */
```
Please restore this comment.

thelovekesh replied:
Upon checking it out, I don't think there is the documentation for `amp_register_polyfills` in `includes/class-amp-theme-support.php`. I think it should have been removed in some other changes.

westonruter replied:
You're right. All instances of this comment should be replaced rather with a reference to `src/Admin/PairedBrowsing.php` which is where the phpdoc is now located.

westonruter commented on src/Admin/ValidationCounts.php:
```diff
@@ -122,12 +123,19 @@ public function enqueue_scripts() {
+	/**
+	 * Whether the current screen is pages inside the AMP Options menu.
+	 */
+	public function is_options_menu() {
```
Maybe change this to `is_dedicated_plugin_screen`

thelovekesh committed:
Update is_options_menu() to is_dedicated_plugin_screen() in ValidationCounts class

westonruter commented on src/Admin/ValidationCounts.php:
```diff
@@ -122,12 +123,19 @@ public function enqueue_scripts() {
-		) {
-			$this->rest_preloader->add_preloaded_path( '/amp/v1/unreviewed-validation-counts' );
-		}
+		);
```
For `is_needed` the conditions may need to be expanded to account for all of these screens and subscreens:

![image](https://user-images.githubusercontent.com/134745/217340220-5504e7a3-3785-47c6-98aa-c701fe77bccc.png)

thelovekesh replied:
Do you want the `ValidationCounts` service to be needed only on the AMP settings menus?

westonruter replied:
It can be on all screens when we core is shipping the expected version of React, but otherwise it can be limited to only the AMP settings pages.

thelovekesh replied:
`ValidationCounts` doesn't have any dependency on React and uses `'wp-api-fetch', 'wp-dom-ready', 'wp-i18n'`(which also doesn't have any dependency on React). So I think we can keep it on all screens.

westonruter replied:
Ah, good point. Nevertheless, we will need to _not_ register the polyfills for this since the polyfills are unconditional now.

westonruter replied:
That is, unless we limit the service to only run on our dedicated screens. But probably simpler to just remove the polyfill registration entirely for this. Since the service is already conditional based on whether it `has_support`, there doesn't seem to be any need for polyfills anyway, right?

thelovekesh replied:
Yes, we can remove polyfills for it entirely as it's conditional based on `has_support`.

thelovekesh replied:
Removed the `amp_register_polyfills` hook from the `ValidationCounts` service.

thelovekesh committed:
Update docblock reference for amp_register_polyfills

westonruter commented on includes/admin/class-amp-template-customizer.php:
```diff
@@ -330,8 +330,12 @@ public function add_customizer_scripts() {
 
-		/** This action is documented in includes/class-amp-theme-support.php */
-		do_action( 'amp_register_polyfills' );
+		if ( ! Services::get( 'dependency_support' )->has_support() ) {
```
Aside: This is essentially re-making the `Polyfills` service conditional which was removed in https://github.com/ampproject/amp-wp/pull/7421

Interestingly, also, there are some cases where this will never be reachable:

https://github.com/ampproject/amp-wp/blob/3c17ea98510c637c671d88810c11e057ac71c5b0/includes/admin/functions.php#L18-L54

thelovekesh replied:
>Aside: This is essentially re-making the Polyfills service conditional which was removed in https://github.com/ampproject/amp-wp/pull/7421

But only for customizers right?

> Interestingly, also, there are some cases where this will never be reachable:

Actually, added it as a safeguard. Should we remove it? What if it surpassed the condition in any case?

westonruter replied:
I'm unsure. Maybe what we should do is just change `amp_init_customizer` to just `return` early if `! Services::get( 'dependency_support' )->has_support()`. 

And then at the same time, we can _remove_ all of the `amp_register_polyfills` actions from this class. What this means is that the custom AMP Customizer integrations won't be available on WP<5.6, but that's effectively the earliest version we support anyway. 

Additionally, Customizer usage is declining.

westonruter replied:
So this means replacing all of the contents of this `if` condition:

https://github.com/ampproject/amp-wp/blob/3c17ea98510c637c671d88810c11e057ac71c5b0/includes/admin/functions.php#L20-L51

With just `return`.

thelovekesh replied:
I think you mean:

```php
if ( ! Services::get( 'dependency_support' )->has_support() ) {
		return;
}
```

We can also remove the `@codeCoverageIgnore` right?

westonruter replied:
Yes, that's right.

westonruter replied:
Or rather, you can make it:

```php
if ( ! Services::get( 'dependency_support' )->has_support() ) {
    return; // @codeCoverageIgnore
}
```

thelovekesh committed:
Update webpack config to bundle @wordpress/dom-ready package with default export library

thelovekesh committed:
Remove amp_register_polyfills hook from ValidationCounts service

thelovekesh committed:
Remove amp customizer init if WP < 5.6

thelovekesh committed:
Remove amp_register_polyfills hook from AMP_Template_Customizer

- We are not using amp customizer on WP < 5.6, so polyfills are not needed. See: https://github.com/ampproject/amp-wp/pull/7458#discussion_r1099147762

thelovekesh committed:
Remove amp_register_polyfills hook from PairedBrowsing

thelovekesh committed:
Remove amp_register_polyfills hook assertions from PairedBrowsingTest

thelovekesh committed:
Skip AMP customizer tests below WP 5.6

thelovekesh committed:
Update test skip message

thelovekesh committed:
Skip tests for PairedBrowsing if WP < 5

westonruter reviewed (approved):
Great work! This is now ready for QA. 🎉
